### ISSUE TITLE
Windows build

### ISSUE THREAD

**Issue by OliverCole** [beakerbrowser/beaker#55]
Multiple failures with build on Windows 10, Node 6.2.2.

[beaker-windows-fail.txt](https://github.com/pfrazee/beaker/files/416590/beaker-windows-fail.txt)

Highlights include:

> Error: no binary available for os:windows

and

> rm -Rf ./node_modules ./app/node_modules; npm i; npm run rebuild
> 
> 'rm' is not recognized as an internal or external command,
> operable program or batch file.

**Comment by pfrazee** [comment]
Adding to the 0.2.0 milestone. Should be ready in ~2 weeks.

**Comment by pmario** [comment]
I did just compile beaker on win10-pro ... It failed quite a bit :/

The following pre requisites are needed: 

 - [VisualStudio Free 2015](https://github.com/electron/electron/blob/master/docs/development/build-instructions-windows.md)
 - python 2.7 [see:](https://github.com/electron-userland/electron-builder/issues/638#issuecomment-237496109)
    - `npm config set python c:/python2.7`  .. did the trick for me. c:/python2.7 is default from install 
 - [node-sodium](https://github.com/mafintosh/node-sodium-prebuilt#windows-install)
   - `npm config set msvs_version 2015`
 - `npm install -g node-gyp` ... optional
 - `npm install -g gulp` ... optional

after I did set the above parameters I did: 

 - `git clone https://github.com/beakerbrowser/beaker.git`
 - `cd beaker`
 - `npm install` ... fails
 - `cd app`
    - `npm rebuild --runtime=electron --target=1.5.0 --disturl=https://atom.io/download/atom-shell --build-from-source`
    - ... there is an error. -> ignore it
    - `gulp build`
    - `cd ..`
 - `npm start`

and it worked \o/

**Comment by pfrazee** [comment]
🎊  you're the first!

**Comment by pfrazee** [comment]
Soon as I can get a win10 box to work with, I'll reproduce your steps and update the readme

**Comment by pmario** [comment]
full: log: http://pastebin.com/wWuWN8U0 ... will expire in 1 month

**Comment by pmario** [comment]
>Soon as I can get a win10 box to work with, I'll reproduce your steps and update the readme

I think, that's not ready for "prime time".. it's super hacky atm :)

It's possible, that others miss some more dependencies, since I already have a dev machine. ... It would be possible, to set up a VM with a vanilla win10 and see what's really needed. ... but that's a pain, since you'll need to make all the win10 updates :/

**Comment by pmario** [comment]
>rm -Rf ./node_modules ./app/node_modules; npm i; npm run rebuild

powershell knows `rm` but it doesn't know `rm -Rf` ... [rimraf](https://www.npmjs.com/package/rimraf) will do the trick.

**Comment by rocketinventor** [comment]
Why is this being run in PowerShell and not Git Bash, Windows 10 Bash, or Cygwin?

**Comment by pmario** [comment]
>Why is this being run in PowerShell and not Git Bash, Windows 10 Bash, or Cygwin?

@rocketinventor What do you mean with "this"?
 - `rm -Rf`  or
 - compiling beaker browser

**Comment by rocketinventor** [comment]
@pmario compiling beaker browser

**Comment by pmario** [comment]
@rocketinventor ... I did try it with git bash, but it failed, so I did try powershell (because it's there). It also failed but I "fiddled" till it worked. 

Windows 10 bash is ubuntu bash so wrong platform. 
windows 10 cmd is "dos" and doesn't even understand `ls`
Cygwin ... would need to install it. So why should I, if the other stuff is already there.

**Comment by rocketinventor** [comment]
cygwin...because it has a terminal, linux binaries, and can generate exe's (I think).
cmd ... use `dir` instead of `ls`. cmd can actually do quite a lot (including scripting) but most of the commands are slightly different.

**Comment by pfrazee** [comment]
It'd be smart to make PowerShell AND one of Git Bash, Windows 10 Bash, or Cygwin work. It's a v1.0 goal for us and I dont have a windows box available, so I haven't put my energy into this yet.

**Comment by pmario** [comment]
TLDR;
I think, before we go on here, it's more important to make the tests run on macOS. see: 292 ... We need a reference for windows and linuy.  ... I don't own a mac.

---------

I always try to test with powershell AND gitbash on windows. In normal cases both seem to work well with nodejs projects. .. At least it is intended by most project creators. ... but as I wrote eg: `rm -Rf` won't work in powershell but does in gitbash. So we'll need some changes to make them both work. ... But that's "low hanging fruit" for `rm`.

The biggest challenges are native node modules. .. They cause some errors atm which stop the install process. ... I could work around the hickups (see my first post), to get beaker browser working, just enough to test something, I was interested in. ... 

And I wanted to let others know, so we can improve the build scripts. IMO the goal should be:

```
npm install
npm start 
```

**Comment by rjsteinert** [comment]
If this is an issue isolated to the fact that Bash is requirement for the build scripts, we could build on a Windows system with Bash :). If the resulting build then requires Bash to continue to be there, then that would limit the number of Windows machines are going to jump in and run Beaker, but at least it would be a start. Wish I had Windows machine to test on.  Any Microsoft folks listening? ;)

[Install the Windows Subsystem for Linux](https://msdn.microsoft.com/en-us/commandline/wsl/install_guide)

**Comment by rocketinventor** [comment]
@rjsteinert Running Beaker on WinBash basically the same thing as running it on Linux. Any build made on it with GCC will only work from within a Linux environment. However, it might be possible to compile a Windows executable using mingw-w64. I have a Windows box and could do basic testing but I don't have enough time to do a lot of debugging if it doesn't work. 

_Sent from my Samsung SM-G950U using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by pfrazee** [comment]
The build tools are not the blocker, it's UTP support.

**Comment by pmario** [comment]
Here's a link to the issue(s) to keep track of: https://github.com/datproject/dat/issues/581

**Comment by pfrazee** [comment]
Someone in the community made a guide for building on windows currently. Still no utp support though: http://awesomeprojectsxyz.blogspot.it/2017/08/compile-beaker-browser-to-run-on.html

**Comment by p1Ggy** [comment]
I'm eagerly waiting a Windows version.

**Comment by pfrazee** [comment]
Another good guide: http://jaredthomson.com/posts/2017/10/15/rotonde-on-windows

**Comment by xoorath** [comment]
thanks @pfrazee, credit still goes to Orfeo Morello's post you linked before. It's where I got most of the steps, just tweaked some.

**Comment by xoorath** [comment]
I've been looking into utp on windows. I made some notes here: [link](https://github.com/mafintosh/utp-native/issues/8)

**Comment by taravancil** [comment]
Not to get y'all too excited, but @goldbuick has made some progress :tada:

https://github.com/mafintosh/utp-native/pull/21

**Comment by ragekit** [comment]
I managed to build on windows with utp native from the mafintosh repo (it does not seem updated in npm yet)
Still had to do @xoorath / Morello little hack with sqlite3 but it worked like a charm
You don't need to install vs studio : you can npm install-g windows-build-tools , thatll take care of python and the windows build tools.

**Comment by pmario** [comment]
>npm install-g windows-build-tools , thatll take care of python and the windows build tools.

If python isn't part of your path variable this didn't work for me. So I needed to set it. 

Using **Powershell in Admin mode** (This worked for me, but depending on your tools, that have been installed already. .. This may vary!)

see: https://www.npmjs.com/package/windows-build-tools

```
npm --add-python-to-path install -g windows-build-tools
```

**Comment by pmario** [comment]
@pfrazee ... Do you accept PRs to improve the windows build process? eg: using `rimraf`[see my other post](https://github.com/beakerbrowser/beaker/issues/55#issuecomment-279551093)

**Comment by pfrazee** [comment]
Yeah feel free to make them. I'm going to spin up a VM or install this coming week, so I'll test out any PRs you make then.

**Comment by pfrazee** [comment]
I was able to get a build going on a win10 VM, including with an installer. Still some todos:

 - [ ] Fix the installer icon
 - [ ] Update all `npm` commands to be portable
 - [ ] Fix the UI in the titlebar (is too ugly for words rn)

**Comment by pfrazee** [comment]
@pmario I handled the script portability, no need to make a PR

**Comment by pfrazee** [comment]
Note, due to https://github.com/npm/npm/issues/17747 you have to use npm4 to build on windows.

**Comment by pmario** [comment]
>Note, due to npm/npm#17747 you have to use npm4 to build on windows.

For me it seems to work with whatever comes with nodejs v6.11.x LTS

**Comment by pfrazee** [comment]
It is done!
